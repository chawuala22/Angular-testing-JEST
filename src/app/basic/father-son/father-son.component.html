<h5>Hijo</h5>

<code> {{client | json}}</code>

<button *ngIf="client" (click)="onChange(5)" data-test="btnUpdate">
    Cambiar ID
</button>

<button *ngIf="client" (click)="onDelete()" data-test="btnDelete">
    Eliminar cliente
</button>