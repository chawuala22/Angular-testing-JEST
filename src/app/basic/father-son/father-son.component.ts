import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss'],
})
export class FatherSonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onUpdateClient = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }
  onChange(id: number) {
    if (!this.client) return;

    // this.client.id = newId no usar

    this.client = {...this.client,id};

    this.onUpdateClient.emit(this.client);
  }
}
