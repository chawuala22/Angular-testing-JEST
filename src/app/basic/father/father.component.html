<h1>Padre</h1>
<button (click)="onSetClient('Miguel')">Cargar cliente</button>

<br /><br />

<code class="mt-2"  >{{ client | json }}</code>

<hr />

<app-father-son
  [client]="client"
  (onDeleteClient)="client = undefined"
  (onUpdateClient)="client= $event"
></app-father-son>
