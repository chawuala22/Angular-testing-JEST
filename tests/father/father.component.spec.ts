import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherComponent } from '../../src/app/basic/father/father.component';
import { FatherSonComponent } from '../../src/app/basic/father-son/father-son.component';
import { By } from '@angular/platform-browser';

describe('FatherComponent', () => {
  let component: FatherComponent;
  let fixture: ComponentFixture<FatherComponent>;
  let compiled : HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherComponent, FatherSonComponent ]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FatherComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe hacer match con snapshot', ()=>{
    expect(compiled).toMatchSnapshot();
  });


  test('Debe establecer el cliente con el nombre indicado', ()=>{
    component.onSetClient('Pedro');
    fixture.detectChanges();

    const codeDiv = compiled.querySelector('.mt-2');
    expect (codeDiv?.textContent).toContain('"name');
    expect (codeDiv?.textContent).toContain('"Pedro');
  });

  test('Debe borrar el cliente si se emite onDeleteClient (hijo)', ()=>{

    component.client={
      id:1,
      name:"Eduardo"
    };
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query( By.directive(FatherSonComponent) );
    const sonComponent: FatherSonComponent = sonDebugElement.componentInstance;
    sonComponent.onDeleteClient.emit();
    expect ( component.client ).toBe(undefined);
  });

  test('Debe actualizar el cliente con onClientUpdate',()=>{

    component.client={
      id:1,
      name:"Eduardo"
    };
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query( By.directive(FatherSonComponent) );
    const sonComponent: FatherSonComponent = sonDebugElement.componentInstance;
    sonComponent.onUpdateClient.emit({id:10, name:'Pedro'});
    expect ( component.client ).toEqual({id:10, name:'Pedro'});
  })

});
