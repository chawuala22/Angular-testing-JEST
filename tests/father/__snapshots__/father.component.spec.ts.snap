// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherComponent debe hacer match con snapshot 1`] = `
<div
  id="root1"
>
  <h1>
    Padre
  </h1>
  <button>
    Cargar cliente
  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
    
  </code>
  <hr />
  <app-father-son>
    <h5>
      Hijo
    </h5>
    <code>
       
    </code>
    
    
  </app-father-son>
</div>
`;
